fe/fe-core/src/main/java/com/starrocks/sql/parser/AstBuilder.java

```diff
@@ -0,0 +1,338 @@
+import com.starrocks.sql.ast.UnionRelation;
+import com.starrocks.sql.ast.ValuesRelation;
+import com.starrocks.sql.optimizer.base.SetQualifier;
+import org.antlr.v4.runtime.ParserRuleContext;
+import org.antlr.v4.runtime.tree.TerminalNode;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static java.util.stream.Collectors.toList;
+
+public class AstBuilder extends StarRocksBaseVisitor<ParseNode> {
+    @Override
+    public ParseNode visitSingleStatement(StarRocksParser.SingleStatementContext context) {
+        return visit(context.statement());
+    }
+
+    // -------------------------------- Statement ------------------------------
+
+    @Override
+    public ParseNode visitQueryStatement(StarRocksParser.QueryStatementContext context) {
+        QueryRelation queryRelation = (QueryRelation) visit(context.query());
+        return new QueryStatement(queryRelation);
+    }
+
+    @Override
+    public ParseNode visitQuery(StarRocksParser.QueryContext context) {
+        QueryRelation queryRelation = (QueryRelation) visit(context.queryNoWith());
+
+        List<CTERelation> withQuery = new ArrayList<>();
+        if (context.with() != null) {
+            withQuery = visit(context.with().namedQuery(), CTERelation.class);
+        }
+        withQuery.forEach(queryRelation::addCTERelation);
+
+        return queryRelation;
+    }
+
+    @Override
+    public ParseNode visitNamedQuery(StarRocksParser.NamedQueryContext context) {
+        Optional<List<Identifier>> columns = Optional.empty();
+        if (context.columnAliases() != null) {
+            columns = Optional.of(visit(context.columnAliases().identifier(), Identifier.class));
+        }
+
+        List<String> columnNames = null;
+        if (columns.isPresent()) {
+            columnNames = columns.get().stream().map(Identifier::getValue).collect(toList());
+        }
+
+        return new CTERelation(
+                ((Identifier) visit(context.name)).getValue(),
+                columnNames,
+                (QueryRelation) visit(context.query()));
+    }
+
+    @Override
+    public ParseNode visitQueryNoWith(StarRocksParser.QueryNoWithContext context) {
+
+        List<OrderByElement> orderByElements = new ArrayList<>();
+        if (context.ORDER() != null) {
+            orderByElements.addAll(visit(context.sortItem(), OrderByElement.class));
+        }
+
+        LimitElement limitElement = null;
+        if (context.limitElement() != null) {
+            limitElement = (LimitElement) visit(context.limitElement());
+        }
+
+        QueryRelation term = (QueryRelation) visit(context.queryTerm());
+        term.setOrderBy(orderByElements);
+        term.setLimitElement(limitElement);
+        return term;
+    }
+
+    @Override
+    public ParseNode visitQuerySpecification(StarRocksParser.QuerySpecificationContext context) {
```

Comment:
Not use `QuerySpecification` again.

Reply:
In the grammar file, QuerySpecification is a standard name for a select statement. For example, in the sqlserver documentation（https://docs.microsoft.com/en-us/dotnet/api/microsoft.sqlserver.transactsql.scriptdom.queryspecification?view=sql-dacfx-150）, it is clearly stated that QuerySpecification means "This class represents the major part of the select statement."
In the mysql grammar file ([querySpecification:](https://github.com/mysql/mysql-workbench/blob/2f35f9034f015cbcd22139a60e1baa2e3e8e795c/library/parsers/grammars/MySQLParser.g4#L1062)) officially given by Oracle, QuerySpecification is also explicitly used. So at the stage of the grammar file, I suggest to use this name as well, which does not conflict with the subsequent selectRelation